### PR TITLE
add post processing for maskrcnn model

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

add post processing for maskrcnn model, including:
(1) convert 28*28 mask to rle mask
(2) convert output format for model inference.

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/978/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala [hunk 1]
```diff
@@ -77,6 +78,34 @@ class RLEMasks(val counts: Array[Int], val height: Int, val width: Int) extends
...
+  override def equals(obj: Any): Boolean = {
+    if (obj == null) {
+      return false
+    }
+    if (!obj.isInstanceOf[RLEMasks]) {
+      return false
+    }
+    val other = obj.asInstanceOf[RLEMasks]
+    if (this.eq(other)) {
+      return true
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala]: is it true？

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+import scala.collection.mutable.ArrayBuffer
+
+object Utils {
+  // box with 4 element (xyxy)
+  def expandBoxes(bbox: Tensor[Float], bboxExpand: Tensor[Float], scale: Float)
+  : Unit = {
+    require(bbox.nElement() == 4 && bboxExpand.nElement() == 4
+      && bbox.dim() == 1 && bboxExpand.dim() == 1,
+      "Box and expected Box should have 4 elements with one dim")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: expanded box ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+import breeze.linalg.{*, dim, max}
+import com.intel.analytics.bigdl.nn.ResizeBilinear
+import com.intel.analytics.bigdl.nn.abstractnn.DataFormat
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.sun.xml.internal.bind.v2.TODO
+
+import scala.collection.mutable.ArrayBuffer
+
+object Utils {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: private to package ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    bboxExpand.setValue(3, x_c + wHalf)
+    bboxExpand.setValue(2, y_c - hHalf)
+    bboxExpand.setValue(4, y_c + hHalf)
+  }
+
+  // mask with three dims (channel, height, wide)
+  def expandMasks(mask: Tensor[Float], padding: Int): (Tensor[Float], Float) = {
+    require(mask.isContiguous(), "Only support contiguous mask")
+
+    val N = mask.size(1)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: what does `N` `M` mean ? also `pad2` is not good naming

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+          padArray(j + padOffset) = maskArray(offset + maskOffset)
+        }
+      }
+    }
+    (paddedMask, scale)
+  }
+
+  // mask and box should be one by one
+  def decodeMaskInImage(mask: Tensor[Float], box: Tensor[Float],
+    thresh: Float = 0.5f, padding : Int = 1, binaryMask: Tensor[Float]): Unit = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: put the parameters with default value to last ones ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+    if (input_height == output_height && input_width == output_width) {
+      output.copy(input)
+      return
+    }
+
+    require(input.isContiguous() && output.isContiguous(),
+      "Only support contiguous tensor for bilinear")
+    val channels = input.size(1)
+    val idata = input.storage().array()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: inputData ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    val odata = output.storage().array()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: outputData ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    val ioffset = input.storageOffset() - 1
+    val ooffset = output.storageOffset() - 1
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: similar as above

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+    val rheight = areaPixelComputeScale(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: similar as above

**Comment by wzhongyuan** [comment]
Some minor comments about naming variables.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala]: yeah, when they equal, it should be true

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok,changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok,changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 4 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok, changed.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok,changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 6 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok,changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala (hunk 7 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/maskrcnn/Utils.scala]: ok,changed

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/982/

**Pull request merged**